File: bracket_ingredients.js
#!/usr/bin/env node
/**
 * Transform ingredient lines in .nt recipe files by wrapping each ingredient
 * (including quantity) in square brackets.
 *
 * Usage: node bracket_ingredients.js static/recepty_0.nt static/recepty_1.nt
 */

import { readFileSync, writeFileSync } from 'fs';

// ---------------------------------------------------------------------------
// Patterns
// ---------------------------------------------------------------------------

const PREP_PARTICIPLE =
    'nakrájen[aáéý]|rozkrájen[aáéý]|nastrouhan[aáéý]|nasekan[aáéý]|' +
    'nastrúhan[aáéý]|umyt[aáéý]|oloupan[aáéý]|rozmixovan[aáéý]|' +
    'přecezen[aáéý]|nadroben[aáéý]|uvařen[aáéý]|upečen[aáéý]|' +
    'orestovan[aáéý]|osmahnut[aáéý]|propláchnut[aáéý]|prolisovan[aáéý]|' +
    'nakrájenou|nastrouhan[eé]|nasekan[eé]';

const PREP_ADVERB = 'najemno|nahrubo|nadrobno';

const IMPERATIVE_VERBS =
    'nastrouhej|oloupej|nakrájej|rozkrájej|nasekej|rozmíchej';

// ---------------------------------------------------------------------------
// findBracketEnd – return the index where the ingredient name/qty ends
// ---------------------------------------------------------------------------

function findBracketEnd(text) {
    const candidates = [];

    function tryPush(re) {
        const m = re.exec(text);
        if (m && m.index > 0) candidates.push(m.index);
    }

    // parenthetical note: wrap before ' ('
    tryPush(/\s*\(/u);

    // "nebo" / "či" alternatives
    tryPush(/\s+nebo\s/u);
    tryPush(/\s+či\s/u);

    // prep-participle followed by na/do/adverb
    tryPush(new RegExp(
        `\\s+(${PREP_PARTICIPLE})\\s+(?:na\\b|do\\b|${PREP_ADVERB})`, 'iu'
    ));

    // standalone celý/celá/celé at end or before (
    tryPush(/\s+cel[áýé]\s*($|\()/u);

    // em-dash
    tryPush(/\s+[–—]\s+/u);

    // hyphen-dash followed by 2+ uppercase letters
    tryPush(/\s+-\s+[A-ZÁČĎÉĚÍŇÓŘŠŤŮÚÝŽ]{2,}/u);

    // imperative verbs
    tryPush(new RegExp(`\\s+(${IMPERATIVE_VERBS})\\b`, 'iu'));

    // prep adverbs
    tryPush(new RegExp(`\\s+(${PREP_ADVERB})\\b`, 'u'));

    return candidates.length > 0 ? Math.min(...candidates) : text.length;
}

// ---------------------------------------------------------------------------
// splitOutsideParens – split at commas that are outside parentheses
// ---------------------------------------------------------------------------

function splitOutsideParens(text) {
    const parts = [];
    let depth = 0;
    let current = '';

    for (let i = 0; i < text.length; i++) {
        const ch = text[i];
        if (ch === '(') {
            depth++;
            current += ch;
        } else if (ch === ')') {
            depth--;
            current += ch;
        } else if (ch === ',' && depth === 0) {
            const lookahead = text.slice(i + 1).trimStart();
            // Don't split if the next part is a quantity suffix
            if (
                lookahead.startsWith('tj.') ||
                lookahead.startsWith('resp.') ||
                lookahead.startsWith('popř.') ||
                lookahead.startsWith('event.') ||
                /^[\d½¼¾]/u.test(lookahead) ||
                lookahead.startsWith('cca ')
            ) {
                current += ch;
            } else {
                parts.push(current.trim());
                current = '';
            }
        } else {
            current += ch;
        }
    }

    if (current.trim()) parts.push(current.trim());
    return parts.filter(p => p.length > 0);
}

// ---------------------------------------------------------------------------
// transformSinglePart – wrap one ingredient token
// ---------------------------------------------------------------------------

function transformSinglePart(text) {
    text = text.trim();
    if (!text) return text;
    if (text.includes('[')) return text;  // already bracketed

    const end = findBracketEnd(text);
    const main = text.slice(0, end).trim();
    const rest = text.slice(end).trim();

    return rest ? `[${main}] ${rest}` : `[${main}]`;
}

// ---------------------------------------------------------------------------
// splitAtInternalDigit – further split parts like "pepř 3 bobkové listy"
// into ["pepř", "3 bobkové listy"] where a bare single-word ingredient
// is immediately followed by a quantity (missing comma in original text).
// Only fires when the word before the digit is a single word with no digits.
// ---------------------------------------------------------------------------

function splitAtInternalDigit(parts) {
    const result = [];
    for (const part of parts) {
        const trimmed = part.trim();
        // Already starts with a quantity — don't touch
        if (/^[\d½¼¾]/u.test(trimmed) || trimmed.startsWith('cca ')) {
            result.push(part);
            continue;
        }
        // Match: single word (no spaces) then space then digit
        const m = /^(\S+)\s+(\d.*)$/u.exec(trimmed);
        if (m) {
            result.push(m[1], m[2]);
        } else {
            result.push(part);
        }
    }
    return result;
}

// ---------------------------------------------------------------------------
// transformIngredientLine – transform "          - some text"
// ---------------------------------------------------------------------------

function transformIngredientLine(line) {
    const m = /^(\s*-\s)(.*)$/u.exec(line);
    if (!m) return line;

    const prefix = m[1];
    const content = m[2];

    if (!content.trim() || content.includes('[')) return line;

    const parts = splitAtInternalDigit(splitOutsideParens(content));

    if (parts.length > 1) {
        return prefix + parts.map(transformSinglePart).join(', ');
    }
    return prefix + transformSinglePart(content);
}

// ---------------------------------------------------------------------------
// processFile – walk lines and transform ingredient sections
// ---------------------------------------------------------------------------

function processFile(content) {
    const lines = content.split('\n');
    const result = [];
    let inIngredient = false;
    let itemIndent = null;

    for (const line of lines) {
        const stripped = line.trim();

        // Detect "        Ingredience:" key
        const headerMatch = /^(\s+)Ingredience:\s*$/.exec(line);
        if (headerMatch) {
            inIngredient = true;
            itemIndent = headerMatch[1].length + 2; // expected indent of list items
            result.push(line);
            continue;
        }

        if (inIngredient) {
            if (!stripped) {
                result.push(line);
                continue;
            }

            if (stripped === '[]') {
                result.push(line);
                inIngredient = false;
                continue;
            }

            const currentIndent = line.length - line.trimStart().length;

            if (currentIndent === itemIndent && stripped.startsWith('- ')) {
                result.push(transformIngredientLine(line));
                continue;
            }

            // Different indent or new key — exit section
            inIngredient = false;
        }

        result.push(line);
    }

    return result.join('\n');
}

// ---------------------------------------------------------------------------
// Entry point
// ---------------------------------------------------------------------------

const files = process.argv.slice(2);
if (files.length === 0) {
    console.error('Usage: node bracket_ingredients.js <file1.nt> [file2.nt ...]');
    process.exit(1);
}

for (const filepath of files) {
    const content = readFileSync(filepath, 'utf8');
    const newContent = processFile(content);
    writeFileSync(filepath, newContent, 'utf8');
    console.log(`Processed: ${filepath}`);
}

File: bracket_ingredients.py
"""
Transform ingredient lines in .nt recipe files by wrapping each ingredient
(including quantity) in square brackets.

Rules:
1. Only processes lines within `Ingredience:` sections
2. Splits comma-separated items (unless before tj./resp./popř. or the next
   part starts with `cca ` or digits — meaning it's a quantity spec for
   the same ingredient)
3. For each part, wraps before: ( nebo či prep-participle+na celý/celá/celé
   em-dash imperative-verbs
4. Otherwise wraps the whole content
"""

import re
import sys


# ---------------------------------------------------------------------------
# Finding the bracket end-position within a single ingredient token
# ---------------------------------------------------------------------------

PREP_PARTICIPLE = (
    r'nakrájen[aáéý]|rozkrájen[aáéý]|nastrouhan[aáéý]|nasekan[aáéý]|'
    r'nastrúhan[aáéý]|umyt[aáéý]|oloupan[aáéý]|rozmixovan[aáéý]|'
    r'přecezen[aáéý]|nadroben[aáéý]|uvařen[aáéý]|upečen[aáéý]|'
    r'orestovan[aáéý]|osmahnut[aáéý]|propláchnut[aáéý]|prolisovan[aáéý]|'
    r'nakrájenou|nastrouhan[eé]|nasekan[eé]'
)

PREP_ADVERB = r'najemno|nahrubo|nadrobno'

IMPERATIVE_VERBS = (
    r'nastrouhej|oloupej|nakrájej|rozkrájej|nasekej|rozmíchej'
)


def find_bracket_end(text: str) -> int:
    """Return the index where the ingredient name/quantity ends."""
    candidates: list[int] = []

    # --- parenthetical note: wrap before ' (' ---------------------------
    m = re.search(r'\s*\(', text)
    if m and m.start() > 0:
        candidates.append(m.start())

    # --- "nebo" alternative ---------------------------------------------
    m = re.search(r'\s+nebo\s', text)
    if m:
        candidates.append(m.start())

    # --- "či" alternative -----------------------------------------------
    m = re.search(r'\s+či\s', text)
    if m:
        candidates.append(m.start())

    # --- prep-participle followed by na/do/adverb -----------------------
    pattern = (
        r'\s+(' + PREP_PARTICIPLE + r')'
        r'\s+(?:na\b|do\b|' + PREP_ADVERB + r')'
    )
    m = re.search(pattern, text, re.IGNORECASE)
    if m and m.start() > 0:
        candidates.append(m.start())

    # --- standalone celý/celá/celé at end or before ( ------------------
    m = re.search(r'\s+(cel[áýé])\s*($|\()', text)
    if m and m.start() > 0:
        candidates.append(m.start())

    # --- em-dash ( – or — ) --------------------------------------------
    m = re.search(r'\s+[–—]\s+', text)
    if m and m.start() > 0:
        candidates.append(m.start())

    # --- regular hyphen-dash followed by 2+ uppercase letters (warning) -
    m = re.search(r'\s+-\s+[A-ZÁČĎÉĚÍŇÓŘŠŤŮÚÝŽ]{2,}', text)
    if m and m.start() > 0:
        candidates.append(m.start())

    # --- imperative verbs embedded as prep instructions -----------------
    m = re.search(r'\s+(' + IMPERATIVE_VERBS + r')\b', text, re.IGNORECASE)
    if m and m.start() > 0:
        candidates.append(m.start())

    # --- prep adverbs (najemno / nahrubo) that appear AFTER content -----
    m = re.search(r'\s+(' + PREP_ADVERB + r')\b', text)
    if m and m.start() > 0:
        candidates.append(m.start())

    return min(candidates) if candidates else len(text)


# ---------------------------------------------------------------------------
# Splitting a line at commas that are outside parentheses
# ---------------------------------------------------------------------------

def split_outside_parens(text: str) -> list[str]:
    """Split text at commas that are outside parentheses.

    Does NOT split when the next token:
      - starts with a digit / ½ / ¼ / ¾ (quantity suffix like ", 5 kuliček")
      - starts with 'cca '
      - starts with abbreviations: tj., resp., popř., event.
    """
    parts: list[str] = []
    depth = 0
    current: list[str] = []
    i = 0
    n = len(text)

    while i < n:
        ch = text[i]
        if ch == '(':
            depth += 1
            current.append(ch)
        elif ch == ')':
            depth -= 1
            current.append(ch)
        elif ch == ',' and depth == 0:
            lookahead = text[i + 1:].lstrip()
            # Don't split if the next part is a quantity suffix
            if (
                lookahead.startswith(('tj.', 'resp.', 'popř.', 'event.'))
                or re.match(r'[\d½¼¾]', lookahead)
                or lookahead.startswith('cca ')
            ):
                current.append(ch)
            else:
                parts.append(''.join(current).strip())
                current = []
        else:
            current.append(ch)
        i += 1

    if current:
        parts.append(''.join(current).strip())

    return [p for p in parts if p]


# ---------------------------------------------------------------------------
# Transforming a single ingredient token
# ---------------------------------------------------------------------------

def transform_single_part(text: str) -> str:
    text = text.strip()
    if not text:
        return text
    # Already has brackets — leave alone
    if '[' in text:
        return text

    end = find_bracket_end(text)
    main = text[:end].strip()
    rest = text[end:].strip()

    if rest:
        return f'[{main}] {rest}'
    return f'[{main}]'


# ---------------------------------------------------------------------------
# Transforming a full ingredient line  "          - some text"
# ---------------------------------------------------------------------------

def transform_ingredient_line(line: str) -> str:
    m = re.match(r'^(\s*-\s)(.*)', line, re.DOTALL)
    if not m:
        return line

    prefix = m.group(1)
    content = m.group(2)

    # Leave empty or already-bracketed lines alone
    if not content.strip() or '[' in content:
        return line

    parts = split_outside_parens(content)

    if len(parts) > 1:
        transformed = [transform_single_part(p) for p in parts]
        return prefix + ', '.join(transformed)
    else:
        return prefix + transform_single_part(content)


# ---------------------------------------------------------------------------
# Processing a whole file
# ---------------------------------------------------------------------------

def process_file(content: str) -> str:
    lines = content.split('\n')
    result: list[str] = []
    in_ingredient = False
    item_indent: int | None = None

    for line in lines:
        stripped = line.strip()

        # Detect "        Ingredience:" key (at any indent level)
        m = re.match(r'^(\s+)Ingredience:\s*$', line)
        if m:
            in_ingredient = True
            item_indent = len(m.group(1)) + 2   # expected indent of list items
            result.append(line)
            continue

        if in_ingredient:
            if not stripped:
                # Blank line — stay in section (could be spacing between items)
                result.append(line)
                continue

            if stripped == '[]':
                # Empty list marker
                result.append(line)
                in_ingredient = False
                continue

            current_indent = len(line) - len(line.lstrip())

            if current_indent == item_indent and stripped.startswith('- '):
                result.append(transform_ingredient_line(line))
                continue

            # Something at a different indent or a new key — exit section
            in_ingredient = False

        result.append(line)

    return '\n'.join(result)


# ---------------------------------------------------------------------------
# Entry point
# ---------------------------------------------------------------------------

def main() -> None:
    for filepath in sys.argv[1:]:
        with open(filepath, 'r', encoding='utf-8') as f:
            content = f.read()

        new_content = process_file(content)

        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(new_content)

        print(f'Processed: {filepath}')


if __name__ == '__main__':
    main()
